src/domain/auth-rebac/services/src/rebac_service_impl.rs

```diff
@@ -30,19 +32,76 @@ use kamu_auth_rebac::{
     SubjectEntityRelationsError,
     UnsetEntityPropertyError,
 };
+use kamu_core::{
+    DatasetLifecycleMessage,
+    DatasetLifecycleMessageCreated,
+    DatasetLifecycleMessageDeleted,
+    DatasetRepository,
+    MESSAGE_PRODUCER_KAMU_CORE_DATASET_SERVICE,
+};
+use messaging_outbox::{
+    MessageConsumer,
+    MessageConsumerMeta,
+    MessageConsumerT,
+    MessageConsumptionDurability,
+};
 use opendatafabric::{AccountID, DatasetID};
 
+use crate::MESSAGE_CONSUMER_KAMU_REBAC_SERVICE;
+
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 pub struct RebacServiceImpl {
     rebac_repo: Arc<dyn RebacRepository>,
+    dataset_repo_reader: Arc<dyn DatasetRepository>,
 }
 
 #[component(pub)]
 #[interface(dyn RebacService)]
+#[interface(dyn MessageConsumer)]
+#[interface(dyn MessageConsumerT<DatasetLifecycleMessage>)]
+#[meta(MessageConsumerMeta {
+    consumer_name: MESSAGE_CONSUMER_KAMU_REBAC_SERVICE,
+    feeding_producers: &[
+        MESSAGE_PRODUCER_KAMU_CORE_DATASET_SERVICE,
+    ],
+    durability: MessageConsumptionDurability::Durable,
+})]
 impl RebacServiceImpl {
-    pub fn new(rebac_repo: Arc<dyn RebacRepository>) -> Self {
-        Self { rebac_repo }
+    pub fn new(
+        rebac_repo: Arc<dyn RebacRepository>,
+        dataset_repo_reader: Arc<dyn DatasetRepository>,
+    ) -> Self {
+        Self {
+            rebac_repo,
+            dataset_repo_reader,
+        }
+    }
+
+    pub async fn handle_dataset_lifecycle_created_message(
+        &self,
+        message: &DatasetLifecycleMessageCreated,
+    ) -> Result<(), InternalError> {
+        let allows = message.dataset_visibility.is_publicly_available();
+        let (name, value) = DatasetPropertyName::allows_public_read(allows);
+
+        self.set_dataset_property(&message.dataset_id, name, &value)
+            .await
+            .map_err(|err| match err {
+                SetEntityPropertyError::Internal(e) => e,
+            })
+    }
+
+    pub async fn handle_dataset_lifecycle_deleted_message(
+        &self,
+        message: &DatasetLifecycleMessageDeleted,
+    ) -> Result<(), InternalError> {
+        self.delete_dataset_properties(&message.dataset_id)
+            .await
+            .map_err(|err| match err {
+                DeleteEntityPropertiesError::NotFound(e) => e.int_err(),
+                DeleteEntityPropertiesError::Internal(e) => e,
+            })
```

Comment:
Can be replaced with `.await.int_err()`.

I think having `DeleteEntityPropertiesError` in the error chain provides valuable information, which you are loosing by extracting the internal error out of the enum.

Reply:
Important comment, I'll pay more attention to things like this, thanks!

Fixed